examples/analysis/IDE.flix

```diff
@@ -1,182 +1,210 @@
+            if (a1 == a2 and b1 == b2)
+                Transformer.NonBotTrans(a1, b1, JoinLattice.leastUpperBound(c1, c2))
+            else if(c1 == c2)
+                if((a2-a1) != 0 and 0==(b1-b2)%(a2-a1))
+                //                // is divisible
+                Transformer.NonBotTrans(a1, b2, JoinLattice.leastUpperBound(Value.Cst(a1*(b1-b2)/(a2-a1)+b1), JoinLattice.leastUpperBound(c1, c2)))
+                //                // is not divisible
+                else Transformer.NonBotTrans(1, 0, Value.Top)
+            else ???: Transformer
     }
+}
 
-    instance JoinLattice[Transformer] {
-        pub def leastUpperBound(t1: Transformer, t2: Transformer): Transformer = match (t1, t2) {
-            case (Transformer.BotTrans, _) => t2
-            case (_, Transformer.BotTrans) => t1
-            case (Transformer.NonBotTrans(a1, b1, c1), Transformer.NonBotTrans(a2, b2, c2)) =>
-                if (a1 == a2 and b1 == b2)
-                    Transformer.NonBotTrans(a1, b1, JoinLattice.leastUpperBound(c1, c2))
-                else if(c1 == c2)
-                    if(0==(b1-b2)%(a2-a1))
-                    //                // is divisible
-                    Transformer.NonBotTrans(a1, b2, JoinLattice.leastUpperBound(Value.Cst(a1*(b1-b2)/(a2-a1)+b1), JoinLattice.leastUpperBound(c1, c2)))
-                    //                // is not divisible
-                    else Transformer.NonBotTrans(1, 0, Value.Top)
-                else ???: Transformer
-        }
+instance MeetLattice[Transformer] {
+    pub def greatestLowerBound(x: Transformer, y: Transformer): Transformer = ???
+}
+
+instance ToString[Transformer] {
+    pub def toString(x: Transformer): String = match x {
+        case BotTrans               => "BotTrans"
+        case NonBotTrans(n, m, v)   => "NonBotTrans(${n}, ${m}, ${v})"
     }
+}
 
-    instance MeetLattice[Transformer] {
-        pub def greatestLowerBound(x: Transformer, y: Transformer): Transformer = ???
+instance Boxable[Transformer] {
+    pub def box(x: Transformer): Boxed = {
+        let value = x as ##java.lang.Object;
+        let compare = (o1, o2) -> Order.compare(o1 as Transformer, o2 as Transformer);
+        let toString = o -> ToString.toString(o as Transformer);
+        BoxedObject(value, compare, toString)
     }
+    pub def unbox(x: Boxed): Transformer = match x {
+        case BoxedObject(v, _, _) => v as Transformer
+        case _                    => ?bug
+    }
+}
 
-    instance ToString[Transformer] {
-        pub def toString(x: Transformer): String = match x {
-            case BotTrans               => "BotTrans"
-            case NonBotTrans(n, m, v)   => "NonBotTrans(${n}, ${m}, ${v})"
-        }
+instance Order[Transformer] {
+    pub def compare(x: Transformer, y: Transformer): Comparison = match (x, y) {
+        case (BotTrans, BotTrans)                   => EqualTo
+        case (NonBotTrans(t1), NonBotTrans(t2))     => t1 <=> t2
```

Comment:
I found this line and it made me think that either something is very wrong, or I don't understand how match statements work in Flix, cause how can this compile, when NonBotTrans is only used with 1 value but the definition contains 3?

Reply:
Good question. I wonder if it is wrong what is done here.

Nevertheless, in this case `t1` and `t2` become bound to triples which are then compared lexicographically. 

Flix allows a lot of freedom this way. For example, `Some(1, 2, 3)` can be written like that, but also like `Some((1, 2, 3))`. Similarly you can match on both `Some(x)` and `Some(x, y, z)`. In both cases, `Some` is really applied to a triple. This is more general (and "better" than Scala) but at the cost of performance.

Reply:
By the way- thanks for the comments and nice to hear from you again. 

How is life? Everything OK?

Reply:
Oh, that's pretty neat! I wasn't aware of that. 

I'm guessing that this order is also there for printing only, because I'm pretty sure it doesn't match the partial order etc.

I'll send you an email.